lib/active_merchant/billing/gateways/alelo.rb

```diff
@@ -0,0 +1,243 @@
+        ensure_credentials(options, false)
+      end
+
+      def encrypt_payload(body, options)
+        credentials = ensure_credentials(options)
+        key = OpenSSL::PKey::RSA.new(Base64.decode64(credentials[:key]))
+        jwk = JOSE::JWK.from_key(key)
+        alg_enc = { 'alg' => 'RSA-OAEP-256', 'enc' => 'A128CBC-HS256' }
+
+        token = JOSE::JWE.block_encrypt(jwk, body.to_json, alg_enc).compact
+
+        encrypted_body = {
+          token: token,
+          uuid: options[:uuid] || SecureRandom.uuid
+        }
+
+        return encrypted_body.to_json, credentials
+      end
+
+      def parse(body)
+        JSON.parse(body, symbolize_names: true)
+      end
+
+      def post_data(params)
+        params.map { |k, v| "#{k}=#{CGI.escape(v.to_s)}" }.join('&')
+      end
+
+      def commit(action, body, options, try_again = true)
+        payload, credentials = encrypt_payload(body, options)
+
+        if options.dig :http, :method
+          payload = body.to_json if options.dig :http, :prevent_encrypt
+          response = parse ssl_request(options[:http][:method], url(action), payload, request_headers(credentials[:access_token]))
+        else
+          response = parse ssl_post(url(action), payload, request_headers(credentials[:access_token]))
+        end
+
+        resp = Response.new(
+          success_from(action, response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          avs_result: AVSResult.new(code: response['some_avs_response_key']),
+          cvv_result: CVVResult.new(response['some_cvv_response_key']),
+          test: test?
+        )
+
+        return resp unless credentials[:multiresp].present?
+
+        multiresp = credentials[:multiresp]
+        # put code to send back GSF back to the merchant if is needed
+        multiresp.process { resp }
+        multiresp
+      rescue ActiveMerchant::ResponseError => e
+        # Retry on a possible expired encryption key
+        if try_again && e.response.code == '401' && options[:encryption_key].present?
+          options.delete(:access_token)
+          options.delete(:encryption_key)
+          commit(action, body, options, false)
+        else
+          res = parse(e.response.body)
+          Response.new(false, res[:messageUser] || res[:error], res, test: test?)
+        end
+      end
+
+      def success_from(action, response)
+        case action
+        when 'capture/transaction/refund'
+          (response[:status] || '') =~ /estornada/i
+        when 'capture/transaction'
+          (response[:status] || '') =~ /confirmada/i
+        else
+          false
+        end
+      end
+
+      def message_from(response)
+        response[:messages] || response[:messageUser]
+      end
+
+      def authorization_from(response)
+        [response[:requestId], response[:authorizationCode]].join('#')
+      end
+
+      def url(action)
+        "#{test? ? test_url : live_url}#{action}"
```

Comment:
I'm confused by the lack of action specification for the test_url...

Reply:
mmm maybe we were blindly following AM template names, the "action" is just the path for the particular end-point needed, maybe "path" or "resource" is more REST oriented ?

Reply:
Oh - no I think the name is ok. What I mean is, why does the live_url have a concatenated action value, but the test_url does not? Is the sandbox just that much different?

Reply:
Oh nevermind - I just can't read 🤦